File: rust/src/poker/hand_value.rs
//! [wiki](https://github.com/lcrocker/ojpoker/wiki/Hand_Value) | Evaluated hand data

use crate::error::{Error,Result};
use crate::cards::*;
use crate::poker::*;

/// [wiki](https://github.com/lcrocker/ojpoker/wiki/Hand_Level) | Class for categories of poker hands
///
/// Poker hands are ranked by first grouping them into categories (which
/// here we call "level"), and then comparing the ranks of the cards
/// within that level to break ties. We use the actual numbers here to
/// index into tables for calculating comparator values.
#[allow(missing_docs)]
#[repr(u8)]
#[derive(Debug, Clone, Copy, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub enum HandLevel {
    #[default]
    None = 0,
    FiveOfAKind = 1,
    StraightFlush = 2,
    Quads = 3,
    FullHouse = 4,
    Flush = 5,
    Straight = 6,
    Trips = 7,
    TwoPair = 8,
    Pair = 9,
    NoPair = 10,
    FourCard = 11, // Badugi
    ThreeCard = 12,
    TwoCard = 13,
    OneCard = 14,
    UnqualifiedFiveOfAKind = 15, // Action Razz
    UnqualifiedQuads = 16,
    UnqualifiedFullHouse = 17,
    UnqualifiedTrips = 18,
    UnqualifiedTwoPair = 19,
    UnqualifiedPair = 20,
    UnqualifiedNoPair = 21,
}

impl std::fmt::Display for HandLevel {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

impl HandLevel {
    /// Convert integer to level
    /// ```rust
    /// use onejoker::prelude::*;
    ///
    /// assert_eq!(HandLevel::FullHouse, HandLevel::from_u8(4));
    /// ```
    pub const fn from_u8(v: u8) -> Self {
        match v {
            1 => HandLevel::FiveOfAKind,
            2 => HandLevel::StraightFlush,
            3 => HandLevel::Quads,
            4 => HandLevel::FullHouse,
            5 => HandLevel::Flush,
            6 => HandLevel::Straight,
            7 => HandLevel::Trips,
            8 => HandLevel::TwoPair,
            9 => HandLevel::Pair,
            10 => HandLevel::NoPair,
            11 => HandLevel::FourCard,
            12 => HandLevel::ThreeCard,
            13 => HandLevel::TwoCard,
            14 => HandLevel::OneCard,
            15 => HandLevel::UnqualifiedFiveOfAKind,
            16 => HandLevel::UnqualifiedQuads,
            17 => HandLevel::UnqualifiedFullHouse,
            18 => HandLevel::UnqualifiedTrips,
            19 => HandLevel::UnqualifiedTwoPair,
            20 => HandLevel::UnqualifiedPair,
            21 => HandLevel::UnqualifiedNoPair,
            _ => HandLevel::None,
        }
    }
}

/// Sentinel value for hand level
pub const HAND_LEVEL_BEST: u8 = 0;
/// Sentinel value for hand level
pub const HAND_LEVEL_WORST: u8 = 0xFF;

/*
 * HandValue
 */

/// The plain integer value used to compare hands--low value wins.
pub type HandValue = u32;

/// For all hand value comparisons, lower is better.
pub const HAND_VALUE_BEST: u32 = 0;
/// Maximum u32
pub const HAND_VALUE_WORST: u32 = 0xFFFF_FFFF;

/*
 * HandDescription
 */

/// [wiki](https://github.com/lcrocker/ojpoker/wiki/Hand_Extras) | Extra information about a hand
#[derive(Debug, Default, Clone, Copy, Hash)]
#[repr(u8)]
pub enum HandExtras {
    /// No extra information
    #[default]
    None = 0,
    /// Cards that each bug represents
    Bugs([Card;2]),
    /// Cards that each wildcard represents
    Wilds([Card;4]),
    /// Bug replacement, up or down?
    Mexican(Card,bool),
}

impl std::fmt::Display for HandExtras {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            HandExtras::None => write!(f, "None"),
            HandExtras::Bugs(b) => write!(f, "Bugs: {} {}", b[0], b[1]),
            HandExtras::Wilds(w) => write!(f, "Wilds: {} {} {} {}", w[0], w[1], w[2], w[3]),
            HandExtras::Mexican(c,b) => write!(f, "Mexican: {} {}", c, b),
        }
    }
}

/// [wiki](https://github.com/lcrocker/ojpoker/wiki/Hand_Description) | Info about an evaluated hand
///
/// Describes all the information about a hand after evaluation.
#[derive(Debug, Clone, Copy)]
#[repr(C)]
pub struct HandDescription {
    /// Copy of the hand under evaluation. Modified during evaluation.
    pub hand: Hand,
    /// Comparison value: lower is better
    pub value: HandValue,
    /// Which game type we are evaluating for
    pub scale: Scale,
    /// Level of the hand
    pub level: HandLevel,
    /// "Meaningful" length
    pub length: u8,
    /// Extra game-specific information about the hand
    pub extras: HandExtras,
}

impl HandDescription {
    /// Create a new HandDescription object from HandValue
    pub fn from_value(hand: &Hand, scale: Scale, value: HandValue) -> Self {
        let mut h = *hand;
        let l = get_ranks_from_value(&mut h, scale, value);

        HandDescription {
            hand: h, scale, value, level: l,
            length: scale.complete_hand() as u8,
            extras: HandExtras::None,
        }
    }

    /// Full english description, e.g. "aces full of kings"
    pub fn full_text(&self) -> String {
        self.scale.full_text(self)
    }

    /// Print relevant part of hand
    pub fn hand_to_string(&self) -> String {
        let mut h = self.hand;
        h.truncate(self.length as usize);
        h.to_string()
    }

    /// Add bug replacement values
    pub fn add_bugs(&mut self, bugs: [Card; 2]) {
        self.extras = HandExtras::Bugs(bugs);
    }
}

impl std::fmt::Display for HandDescription {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} [{}] = {}", self.scale.name(), self.hand, self.value)
    }
}

impl Default for HandDescription {
    fn default() -> Self {
        HandDescription {
            hand: Hand::new(DeckType::English),
            scale: Scale::HighHand,
            value: HAND_VALUE_WORST,
            level: HandLevel::None,
            length: 0,
            extras: HandExtras::None,
        }
    }
}

// Re-order the given hand according to the ranks decoded from the value
fn get_ranks_from_value(h: &mut Hand, g: Scale, value: HandValue)
-> HandLevel {
    assert!(h.len() > 1 && h.len() <= 9);

    let v = value;
    let mut r: Vec<Rank> = Vec::new();
    let lvl = g.level_from_value(0xFF & (v >> 20));

    let digits =
    if g == Scale::Badugi || g == Scale::Badeucy {
        5 - g.value_from_level(lvl) as usize
    } else if h.len() < g.complete_hand() {
        h.len()
    } else {
        g.complete_hand()
    };

    for i in (5-digits)..5 {
        let mut d = 0xF & (v >> (4 * (4 - i)));
        if ! g.low_hands() { d ^= 0xF; }
        r.push(Rank::from_u8(d as u8));
    }

    if h.len() > 5 && (lvl == HandLevel::Flush || lvl == HandLevel::StraightFlush) {
        let mut counts = [0; 4];
        let mut flush_suit = Suit::None;

        // This code may fail for 10-card hands with two flushes
        for i in 0..h.len() {
            counts[h[i].suit() as usize - 1] += 1;
            if counts[h[i].suit() as usize - 1] > 4 {
                flush_suit = h[i].suit();
                break;
            }
        }
        for i in 0..r.len() {
            h[i] = Card::from_rank_suit(r[i], flush_suit);
        }
        return lvl;
    }
    for i in 0..r.len() {
        for j in i..h.len() {
            if h[j].rank() == r[i] {
                if i == j { continue; }
                if h[i].rank() != r[i] || h[j] > h[i] {
                    h[..].swap(i, j);
                }
            }
        }
    }
    lvl
}

/// [wiki](https://github.com/lcrocker/ojpoker/wiki/Hand_Description_Builder) | Builder for hand description
///
/// Builder for new HandDescription objects. Does some error checking while allowing
/// construction im multiple ways.
/// ```rust
/// use onejoker::prelude::*;
/// use onejoker::poker::HandDescriptionBuilder;
///
/// let hand = Hand::new(DeckType::English).init(hand!("As","Js","Ts","7s","5s"));
/// let v = HandDescriptionBuilder::new(&hand, Scale::HighHand)
///    .with_level(HandLevel::Flush)
///    .with_default_value().complete();
/// ```
pub struct HandDescriptionBuilder {
    hand: Hand,
    scale: Scale,
    level: Option<HandLevel>,
    value: Option<HandValue>,
    length: u8,
    extras: HandExtras,
}

impl HandDescriptionBuilder {
    /// Begin creating a new HandDescription object
    pub fn new(hand: &Hand, scale: Scale) -> Self {
        debug_assert!(scale.low_aces() == hand.deck_type().low_aces());
        debug_assert!(hand.len() > 1 && hand.len() <= 13);

        HandDescriptionBuilder {
            hand: *hand, scale, level: None, value: None,
            length: scale.complete_hand() as u8,
            extras: HandExtras::None,
        }
    }

    /// Set the level of the hand
    pub fn with_level(mut self, level: HandLevel) -> Self {
        debug_assert!(self.level.is_none());
        debug_assert!(HAND_LEVEL_WORST as u32 != self.scale.value_from_level(level));

        self.level = Some(level);
        self
    }

    /// Set the value of the hand to a custom number
    pub fn with_value(mut self, value: HandValue) -> Self {
        debug_assert!(self.value.is_none());

        self.value = Some(value);
        self
    }


    /// Set the value of the hand using the default calculation
    pub fn with_default_value(mut self) -> Self {
        debug_assert!(self.value.is_none());
        debug_assert!(self.level.is_some());

        self.value = Some(ojp_default_hand_value(&self.hand,
            self.scale, self.level.unwrap()));
        self
    }

    /// Set "meaningful" length for display
    pub fn truncate(mut self, l: usize) -> Self {
        self.length = l as u8;
        self
    }

    /// Finish building the HandDescription object
    pub fn complete(self) -> Result<HandDescription> {
        if self.level.is_none() {
            return Err(Error::BadDescription("level not set".into()));
        }
        if self.value.is_none() {
            return Err(Error::BadDescription("value not set".into()));
        }
        Ok(
            HandDescription {
                hand: self.hand,
                scale: self.scale,
                level: self.level.expect("can't happen"),
                value: self.value.expect("can't happen"),
                length: self.length,
                extras: self.extras,
            }
        )
    }
}

impl PartialEq for HandDescription {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}
impl Eq for HandDescription {}

impl PartialOrd for HandDescription {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.value.cmp(&other.value))
    }
}

impl Ord for HandDescription {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.value.cmp(&other.value)
    }
}

impl std::hash::Hash for HandDescription {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.value.hash(state);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cmp::{PartialOrd, PartialEq, Eq, Ord};
    use std::marker::{Sized, Send, Sync, Unpin};
    use std::fmt::{Debug, Display};

    fn has_traits<T: Debug + Display + PartialOrd + PartialEq + Eq + Ord + Clone + Copy +
        std::hash::Hash + std::default::Default + Sized + Send + Sync + Unpin>() {}

    #[test]
    fn test_hand_evaluator_high() -> Result<()> {
        has_traits::<HandLevel>();
        // has_traits::<HandExtras>(); No meaningful Ord
        has_traits::<HandDescription>();
        Ok(())
    }
}

File: dart/lib/src/poker/hand_value.dart
// ignore_for_file: constant_identifier_names
import '../cards/cards.dart';
import '../poker/high_hand.dart';

abstract class HandLevelInterface {
  int get index;
  HandLevelInterface fromIndex(int x);
  HandLevelInterface get best;
  HandLevelInterface get worst;
}

enum HandLevelHigh implements HandLevelInterface {
  None, // 0
  FiveOfAKind, // 1
  StraightFlush, // 2
  Quads, // 3
  FullHouse, // 4
  Flush, // 5
  Straight, // 6
  Trips, // 7
  TwoPair, // 8
  Pair, // 9
  NoPair; // 10

  @override
  HandLevelHigh fromIndex(int x) {
    return values[x];
  }

  @override
  HandLevelHigh get best {
    return FiveOfAKind;
  }

  @override
  HandLevelHigh get worst {
    return NoPair;
  }
}

typedef HandLevelPaiGow = HandLevelHigh;

enum HandLevelStripped implements HandLevelInterface {
  None, // 0
  FiveOfAKind, // 1
  StraightFlush, // 2
  Quads, // 3
  Flush, // 4
  FullHouse, // 5
  Straight, // 6
  Trips, // 7
  TwoPair, // 8
  Pair, // 9
  NoPair; // 10

  @override
  HandLevelStripped fromIndex(int x) {
    return values[x];
  }

  @override
  HandLevelStripped get best {
    return FiveOfAKind;
  }

  @override
  HandLevelStripped get worst {
    return NoPair;
  }
}

typedef HandLevelManilla = HandLevelStripped;
typedef HandLevelMexican = HandLevelStripped;

enum HandLevelAceToFive implements HandLevelInterface {
  None, // 0
  NoPair, // 1
  Pair, // 2
  TwoPair, // 3
  Trips, // 4
  FullHouse, // 5
  Quads; // 6

  @override
  HandLevelAceToFive fromIndex(int x) {
    return values[x];
  }

  @override
  HandLevelAceToFive get best {
    return NoPair;
  }

  @override
  HandLevelAceToFive get worst {
    return Quads;
  }
}

enum HandLevelDeuceToSeven implements HandLevelInterface {
  None, // 0
  NoPair, // 1
  Pair, // 2
  TwoPair, // 3
  Trips, // 4
  Straight, // 5
  Flush, // 6
  FullHouse, // 7
  Quads, // 8
  StraightFlush; // 9

  @override
  HandLevelDeuceToSeven fromIndex(int x) {
    return values[x];
  }

  @override
  HandLevelDeuceToSeven get best {
    return NoPair;
  }

  @override
  HandLevelDeuceToSeven get worst {
    return Quads;
  }
}

typedef HandLevelAceToSix = HandLevelDeuceToSeven;

enum HandLevelBadugi implements HandLevelInterface {
  None, // 0
  FourCard, // 1
  ThreeCard, // 2
  TwoCard, // 3
  OneCard; // 4

  @override
  HandLevelBadugi fromIndex(int x) {
    return values[x];
  }

  @override
  HandLevelBadugi get best {
    return FourCard;
  }

  @override
  HandLevelBadugi get worst {
    return OneCard;
  }
}

enum HandLevelActionRazz implements HandLevelInterface {
  None, // 0
  QualifiedNoPair, // 1
  QualifiedPair, // 2
  QualifiedTwoPair, // 3
  QualifiedTrips, // 4
  QualifiedFullHouse, // 5
  QualifiedQuads, // 6
  UnqualifiedNoPair, // 7
  UnqualifiedPair, // 8
  UnqualifiedTwoPair, // 9
  UnqualifiedTrips, // 10
  UnqualifiedFullHouse, // 11
  UnqualifiedQuads; // 12

  @override
  HandLevelActionRazz fromIndex(int x) {
    return values[x];
  }

  @override
  HandLevelActionRazz get best {
    return QualifiedNoPair;
  }

  @override
  HandLevelActionRazz get worst {
    return UnqualifiedQuads;
  }
}

/// Default function for assigning a numeric value to the [Hand] such that
/// lower number equals better hand.
///
/// Start with just the level times 10 million.
/// Then within each level, use `PositionalHash.u32csr()` to
/// express the ranks as an n-digit number in base 16 (which should be less
/// than 10000000). That number will be high for better hands within the
/// level, so we negate it.
int ojHighHandValueFunction(HandLevelInterface level, List<Rank> ranks) {
  return 10000000 * level.index - ojhPositional32cr(ranks);
}

int ojLowHandValueFunction(HandLevelInterface level, List<Rank> ranks) {
  return 10000000 * level.index + ojhPositional32cr(ranks);
}

/// All the information resulting from the evaluation of a poker hand.
///
/// This is used for comparing hands to determine a winner, and also for
/// displaying the hand appropriately.
/// Poker hands are typically evaluated into one of a small set of
/// categories we call "levels" (e.g. Pair, Flush, Straight) and then
/// within that category, the ranks of the cards are compared.
/// All the numbers here are such that lower number means better hand.
/// {@category poker}
class HandValue<L extends HandLevelInterface>
    implements Comparable<HandValue<L>> {
  final L level;
  final List<Rank> ranks;
  final int value;

  /// Constructor for internal use
  HandValue(this.level, this.ranks, this.value);

  // Another ugly hack to work around Dart's broken generics.
  // Should be able to just use L.worst here.
  HandValue.worst()
      : level = HandLevelActionRazz.UnqualifiedQuads as L,
        ranks = [],
        value = 0x7FFFFFFFFFFFFFFF;

  /// Implement == for [HandValue]
  @override
  bool operator ==(Object other) {
    if (other is HandValue) {
      return value == other.value;
    }
    return false;
  }

  /// Overriding == requires overriding hashCode as well
  @override
  int get hashCode {
    return ojHighHandValueFunction(level, ranks);
  }

  @override
  int compareTo(HandValue<L> other) {
    return value - other.value;
  }

  bool operator <(HandValue<L> other) {
    return value < other.value;
  }

  bool operator <=(HandValue<L> other) {
    return value <= other.value;
  }

  bool operator >(HandValue<L> other) {
    return value > other.value;
  }

  bool operator >=(HandValue<L> other) {
    return value >= other.value;
  }

  /// Describe the [Hand] (we expect every implementor to override this).
  fullName() {
    return "$level $value";
  }

  /// Re-order the [Card]s for value-appropriate display
  /// (e.g. the hand "5h3cAc3h3d" will display as "3h3d3cAc5h").
  Hand orderedForDisplay(Hand h) {
    List<Card> hIn = h.toList();
    assert(hIn.length == 5);
    assert(ranks.length == 5);
    Hand hOut = h.copyFromIter([]);

    for (int i = 0; i < ranks.length; i += 1) {
      Rank r = ranks[i];
      Card found = Card.None;
      int fIndex = -1;

      for (int j = 0; j < hIn.length; j += 1) {
        if (hIn[j].rank == r && hIn[j] > found) {
          found = hIn[j];
          fIndex = j;
        }
      }
      assert(fIndex >= 0);
      hIn.removeAt(fIndex);
      hOut.push(found);
    }
    assert(hOut.length == 5);
    return hOut;
  }

  /// Implementors will override this with something short for debugging.
  @override
  toString() {
    return fullName();
  }
}

typedef HandValueFactory<V> = V Function();

/// Base class for poker hand evaluation.
///
/// Provides a framework for the subclasses and provides some default
/// implementations.
/// {@category poker}
abstract class HandEvaluator<V extends HandValue> {
  // This abomination is to work around the problem that Dart generics
  // are broken, and do not allow the creation of a generic object in
  // any sensible way.
  HandValueFactory<V> get worstHandValue {
    return () => HandValue.worst() as V;
  }
  int completeHand() { return 5; } 

  V referenceEvaluator(Hand h);

  V partialEvaluator(Hand h) {
    return referenceEvaluator(h);
  }

  V lookupEvaluator(Hand h) {
    return referenceEvaluator(h);
  }

  int fastValue(Hand h) {
    return referenceEvaluator(h).value;
  }

  V valueOf(Hand h) {
    if (h.length < completeHand()) {
      return partialEvaluator(h);
    } else if (h.length == completeHand()) {
      return lookupEvaluator(h);
    } else {
      return bestOf(h);
    }
  }

  V bestOf(Hand h) {
    V best = worstHandValue();

    for (Hand sub in h.combinations(completeHand())) {
      V subV = lookupEvaluator(sub);
      if (subV.value < best.value) {
        best = subV;
      }
    }
    return best;
  }
}

/// Return a [HandEvaluator] by name.
/// {@category poker}
HandEvaluator pokerEvaluator(String name) {
  switch (name) {
    case "standard":
    case "default":
    case "high":
      return HandEvaluatorHigh();
    default:
      throw ArgumentError('Unknown poker evaluator: $name');
  }
}
